### PR TITLE
Unified handling of operator<<(std::ostream, ...), to_string() and print()

### LINKED ISSUE #716: print vs to_string vs operator<<
I feel like we should decide how we want to get a string representation of our classes. Currently, we do all types of things:

```
void AbstractLQPNode::print(std::ostream& out);
std::string to_string(const AllParameterVariant& x);
std::ostream& operator<<(std::ostream& os, const LQPColumnReference& column_reference);
```

I like `to_string` because it is easy to use in tests: `EXPECT_EQ(column_expression1->to_string(), "ColumnID #1 * 2");`
`operator<<` is nice, too, because that's what everyone is used to.
For `print`, I can't come up with something particularly good about it.

### PR DESCRIPTION
**Pull request by mrzzzrm**
Fix #716

Every relevant enum/class/struct is now print-able to streams like so: `std::cout << data_type << lqp_node;`. Previously this was different for each data type: `std::cout << data_type_to_string.left.at(data_type); lqp_node.print(std::cout);`

### CODE REVIEW

File: src/lib/types.hpp [hunk 1]
```diff
@@ -13,6 +8,13 @@
 #include <tuple>
 #include <vector>
 
+#include "tbb/concurrent_vector.h"
```
Comment (mrks) [inline src/lib/types.hpp]: ?

File: src/benchmarklib/benchmark_runner.cpp [hunk 2]
```diff
@@ -271,7 +271,7 @@ void BenchmarkRunner::_benchmark_individual_queries() {
     const auto duration_seconds = static_cast<float>(result.duration_ns) / 1'000'000'000;
     const auto items_per_second = static_cast<float>(result.num_iterations) / duration_seconds;
 
-    std::cout << "  -> Executed " << result.num_iterations << " times in " << duration_seconds << " seconds ("
+    std::cout << "  -> Executed " << result.num_iterations.load() << " times in " << duration_seconds << " seconds ("
```
Comment (mrks) [inline src/benchmarklib/benchmark_runner.cpp]: Why did this become necessary?

File: src/lib/constant_mappings.cpp [hunk 3]
```diff
@@ -97,7 +52,29 @@ const boost::bimap<VectorCompressionType, std::string> vector_compression_type_t
         {VectorCompressionType::SimdBp128, "SIMD-BP128"},
     });
 
-const boost::bimap<TableType, std::string> table_type_to_string =
-    make_bimap<TableType, std::string>({{TableType::Data, "Data"}, {TableType::References, "References"}});
+std::ostream& operator<<(std::ostream& stream, AggregateFunction aggregate_function) {
+  stream << aggregate_function_to_string.left.at(aggregate_function);
+  return stream;
```
Comment (mrks) [inline src/lib/constant_mappings.cpp]: Shouldn't `return stream << ` work, too?

File: src/lib/expression/between_expression.cpp [hunk 4]
```diff
@@ -29,8 +29,7 @@ std::shared_ptr<AbstractExpression> BetweenExpression::deep_copy() const {
 
 std::string BetweenExpression::as_column_name() const {
   std::stringstream stream;
-  stream << _enclose_argument_as_column_name(*value()) << " "
-         << predicate_condition_to_string.left.at(predicate_condition) << " "
+  stream << _enclose_argument_as_column_name(*value()) << " " << predicate_condition << " "
```
Comment (mrks) [inline src/lib/expression/between_expression.cpp]: Can we also `std::cout << *abstract_expression` now?

**Comment by mrks** [comment]
I like this a lot :)

File: src/lib/types.hpp (hunk 1 above)
Comment (mrzzzrm) [reply src/lib/types.hpp]: File uses `tbb::concurrent_vector` - apparently it was transitively included before....

File: src/benchmarklib/benchmark_runner.cpp (hunk 2 above)
Comment (mrzzzrm) [reply src/benchmarklib/benchmark_runner.cpp]: There is no `operator<<(std::ostream&, std::atomic<int>)`, so this previously used `operator<<(..., int)` via [implicit conversion](https://en.cppreference.com/w/cpp/atomic/atomic/operator_T). With the introduction of operator<< overloads for STRONG_TYPEDEFs, there are multiple candidate `operator<<` now - those for the primitive data types and those for the strong typedefs. Calling load() removes the ambiguity.

File: src/lib/expression/between_expression.cpp (hunk 4 above)
Comment (mrzzzrm) [reply src/lib/expression/between_expression.cpp]: yes, but in this case `_enclose_argument_as_column_name` is necessary to ensure parentheses are printed to make the printed expression non-ambiguous

File: src/lib/constant_mappings.cpp (hunk 3 above)
Comment (mrzzzrm) [reply src/lib/constant_mappings.cpp]: ehrm yes. good point!

**Comment by mrzzzrm** [comment]
> I only scrolled over it. Did I miss anything in here that requires a proper review?

There shouldn't be hidden traps or anything. STRONG_TYPEDEF includes generating a operator<<, but that's the most significant thing that comes to mind.

File: src/lib/types.hpp (hunk 1 above)
Comment (mrks) [reply src/lib/types.hpp]: No, it was also included before. You switched it from <> to "".

Comment (mrzzzrm) [reply src/lib/types.hpp]: probably because I like using <> for stdlib an "" for everything else... Want me to revert it?

Comment (mrks) [reply src/lib/types.hpp]: "" looks in the local directory before it looks in the system files. Thus, all system headers should use <>

Comment (mrzzzrm) [reply src/lib/types.hpp]: a right. i keep on forgetting we do not ship boost in-repository.

**Pull request merged**